### ISSUE TITLE
QGIS support editing attribute tables

### ISSUE THREAD

**Issue by visr** [Deltares/Ribasim#841]
**What**
In QGIS, only the records in the attribute tables of Node and Edge can be edited or deleted, but not any of the others.

**Why**
Without this it becomes nearly impossible to create a model, since you can create records but not edit or delete them. It is impossible to change parameters in an existing model.

**How**
QGIS requires a primary key to enable these features (https://github.com/qgis/QGIS/issues/43664). We have this only for the Node and Edge tables. I confirmed this by setting the primary key for only the "Basin / state" table with the following patch of `input_base.py`:

```python
        if table == "Basin / state":
            dtype = {"node_id": "INTEGER PRIMARY KEY"}
        else:
            dtype = {}

        assert self.df is not None
        with closing(connect(temp_path)) as connection:
            self.df.to_sql(
                table,
                connection,
                index=False,
                if_exists="replace",
                dtype=dtype,
            )
```

I suppose we could define a primary key for each data table. For some it would be just `node_id`, for others you need a combination of columns for it to be unique, like (`node_id`, `control_state`), or (`subgrid_id`, `node_id`).

It should be noted that Arrow tables cannot be edited at all in QGIS, but that is less of an issue since they are more used for long timeseries, which you probably don't want to edit by hand anyway.

**Additional context**
Here you can see the desired behavior on "Basin / state" due to the above patch:
![image](https://github.com/Deltares/Ribasim/assets/4471859/9127514d-1b35-4810-84d5-68c644e8ae64)
And here you can see the current behavior, where you can only edit cells for records that you are newly creating:
![image](https://github.com/Deltares/Ribasim/assets/4471859/199b5124-92b6-49c6-a259-2acdf90ead9c)

**Comment by visr** [comment]
> For some it would be just node_id, for others you need a combination of columns for it to be unique, like (node_id, control_state), or (subgrid_id, node_id).

Pandas unfortunately doesn't support setting multiple columns to form a composite primary key. And SQLite doesn't support altering the table to set this manually after the fact; we'd have to create a new table with the right primary key and move the data there.

@deltamarnix mentioned another option would be to add a column to act as a primary key. That might be easier to get to work, perhaps using [autoinc](https://www.sqlite.org/autoinc.html). Though we'd have another column that needs to somehow be incorporated into the datamodel. Setting existing columns as primary key also has the benefit of preventing entering duplicate records.

**Comment by visr** [comment]
If tables are created from QGIS, it contains `fid`, which also solves this issue:

![image](https://github.com/Deltares/Ribasim/assets/4471859/d63ba791-d2f2-407f-9183-744d566f8c6b)

**Comment by deltamarnix** [comment]
Agreed with @visr that we will try the following approach:

- [ ] Add a `meta_row_id` as primary key to all the tables (not Edge or Node). Also to comply with this future feature: #864.
- [ ] Hide the `meta_row_id` columns via the plugin python API on loading of the GeoPackage file. https://gis.stackexchange.com/a/72450

**Comment by visr** [comment]
Composite primary keys are not an option since the GeoPackage spec explicitly requires a integer column as primary key:

> A table SHALL have a primary key column of type INTEGER and that column SHALL act as a rowid alias.

From requirement 119 in http://www.geopackage.org/spec131/#attributes_user_tables.

**Comment by deltamarnix** [comment]
I have tried hiding the fid column, but I found a bug trying so.

https://github.com/qgis/QGIS/issues/55533

It is possible to hide the columns when we load the package from the toml file, but newly added layers that also get an `fid` have wrong column indices. So hiding it would hide a column that is actually important to use.

As long as this bug is not resolved, I would just keep the `fid` column visible.
